File: 0160-intersection-of-two-linked-lists/0160-intersection-of-two-linked-lists.rb
# Definition for singly-linked list.
# class ListNode
#   attr_accessor :val, :next
#   def initialize(val)
#     @val = val
#     @next = nil
#   end
# end

# @param {ListNode} headA
# @param {ListNode} headB
# @return {ListNode}
def getIntersectionNode(headA, headB)
  node1, node2 = headA, headB
  while node1 != node2
    node1 = node1 ? node1.next : headB
    node2 = node2 ? node2.next : headA
  end
  node1
end

File: 0160-intersection-of-two-linked-lists/0160-intersection-of-two-linked-lists.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *   this.val = val;
 *   this.next = null;
 * }
 */

/**
 * @param {ListNode} headA
 * @param {ListNode} headB
 * @return {ListNode}
 */
var getIntersectionNode = function(headA, headB) {
  let [lenA, lenB] = [0, 0];
  let [currA, currB] = [headA, headB];
  while (currA) {
    lenA++;
    currA = currA.next;
  }
  while (currB) {
    lenB++;
    currB = currB.next;
  }
  console.log(lenA, lenB);
  [currA, currB] = [headA, headB];
  if (lenA > lenB) {
    for (let i = lenA - lenB; i > 0; i--) currA = currA.next;
  } else if (lenB > lenA) {
    for (let i = lenB - lenA; i > 0; i--) currB = currB.next;
  }
  while (currA) {
    if (currA === currB) break;
    currA = currA.next;
    currB = currB.next;
  }
  return currA;
};